File: CS/Grid MVVM Filtering/FilteredDataSupport.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.ComponentModel;
using System.Windows;
using System.Collections;
using DevExpress.Xpf.Grid;
using DevExpress.Data.Filtering.Helpers;

namespace Grid_MVVM_Filtering {
    public class FilteredDataSupport {
        public static readonly DependencyProperty VisibleDataProperty =
            DependencyProperty.RegisterAttached("VisibleData", typeof(IList), typeof(FilteredDataSupport), new PropertyMetadata(onDataChanged));

        public static void SetVisibleData(UIElement element, IList value) {
            element.SetValue(VisibleDataProperty, value);
        }
        public static IList GetVisibleData(UIElement element) {
            return (IList)element.GetValue(VisibleDataProperty);
        }

        static void OnCustomRowFilter(object sender, RowFilterEventArgs e) {
            if(e.ListSourceRowIndex != 0)
                return;
            ChangeVisibleData(sender as GridControl);
        }

        static void OnFilterChanged(object sender, RoutedEventArgs e) {
            GridControl grid = sender as GridControl;
            if(grid.IsFilterEnabled == true)
                ChangeVisibleData(grid);
        }

        static void ChangeVisibleData(GridControl grid) {
            var res = grid.DataController.GetAllFilteredAndSortedRows().Cast<object>();
            var visibleData = grid.GetValue(VisibleDataProperty) as IEnumerable<object>;
            var excludedData = grid.GetValue(ExcludedDataProperty) as IEnumerable<object>;
            if(excludedData != null) {
                res = res.Except(excludedData);
            }
            (visibleData as IList).Clear();
            foreach(object item in res) {
                (visibleData as IList).Add(item);
            }
        }

        private static Type GetItemType(IEnumerable data) {
            foreach(object item in data)
                return item.GetType();
            return null;
        }

        public static readonly DependencyProperty ExcludedDataProperty =
            DependencyProperty.RegisterAttached("ExcludedData", typeof(IList), typeof(FilteredDataSupport), new PropertyMetadata(null, onDataChanged, coerceExcludedData));

        public static void SetExcludedData(UIElement element, IList value) {
            element.SetValue(ExcludedDataProperty, value);
        }
        public static IList GetExcludedData(UIElement element) {
            return (IList)element.GetValue(ExcludedDataProperty);
        }

        private static void onDataChanged(DependencyObject sender, DependencyPropertyChangedEventArgs e) {
            GridControl grid = sender as GridControl;
            if(grid == null)
                return;
            if(e.OldValue == null && e.NewValue != null) {
                if(e.Property == ExcludedDataProperty)
                    grid.CustomRowFilter += excludeOnCustomRowFilter;
                else if(e.Property == VisibleDataProperty) {
                    grid.CustomRowFilter += OnCustomRowFilter;
                    grid.FilterChanged += OnFilterChanged;
                }
            } else if(e.OldValue != null && e.NewValue == null) {
                if(e.Property == ExcludedDataProperty)
                    grid.CustomRowFilter -= excludeOnCustomRowFilter;
                else if(e.Property == VisibleDataProperty) {
                    grid.CustomRowFilter -= OnCustomRowFilter;
                    grid.FilterChanged -= OnFilterChanged;
                }
            }
            grid.RefreshData();
        }

        private static object coerceExcludedData(DependencyObject sender, object data) {
            GridControl grid = sender as GridControl;
            grid.RefreshData();
            return data;
        }

        static void excludeOnCustomRowFilter(object sender, RowFilterEventArgs e) {
            GridControl grid = sender as GridControl;
            IList data = grid.ItemsSource as IList;
            IList excludedData = grid.GetValue(ExcludedDataProperty) as IList;
            e.Visible = !excludedData.Contains(data[e.ListSourceRowIndex]);
            e.Handled = !e.Visible;
        }
    }
}

File: VB/Grid MVVM Filtering/FilteredDataSupport.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.ComponentModel
Imports System.Windows
Imports System.Collections
Imports DevExpress.Xpf.Grid
Imports DevExpress.Data.Filtering.Helpers

Namespace Grid_MVVM_Filtering
	Public Class FilteredDataSupport
        Public Shared ReadOnly VisibleDataProperty As DependencyProperty = DependencyProperty.RegisterAttached("VisibleData", GetType(IList), GetType(FilteredDataSupport), New PropertyMetadata(AddressOf onDataChanged))

		Public Shared Sub SetVisibleData(ByVal element As UIElement, ByVal value As IList)
			element.SetValue(VisibleDataProperty, value)
		End Sub
		Public Shared Function GetVisibleData(ByVal element As UIElement) As IList
			Return CType(element.GetValue(VisibleDataProperty), IList)
		End Function

		Private Shared Sub OnCustomRowFilter(ByVal sender As Object, ByVal e As RowFilterEventArgs)
			If e.ListSourceRowIndex <> 0 Then
				Return
			End If
			ChangeVisibleData(TryCast(sender, GridControl))
		End Sub

		Private Shared Sub OnFilterChanged(ByVal sender As Object, ByVal e As RoutedEventArgs)
			Dim grid As GridControl = TryCast(sender, GridControl)
			If grid.IsFilterEnabled = True Then
				ChangeVisibleData(grid)
			End If
		End Sub

		Private Shared Sub ChangeVisibleData(ByVal grid As GridControl)
			Dim res = grid.DataController.GetAllFilteredAndSortedRows().Cast(Of Object)()
			Dim visibleData = TryCast(grid.GetValue(VisibleDataProperty), IEnumerable(Of Object))
			Dim excludedData = TryCast(grid.GetValue(ExcludedDataProperty), IEnumerable(Of Object))
			If excludedData IsNot Nothing Then
				res = res.Except(excludedData)
			End If
			TryCast(visibleData, IList).Clear()
			For Each item As Object In res
				TryCast(visibleData, IList).Add(item)
			Next item
		End Sub

		Private Shared Function GetItemType(ByVal data As IEnumerable) As Type
			For Each item As Object In data
				Return item.GetType()
			Next item
			Return Nothing
		End Function

        Public Shared ReadOnly ExcludedDataProperty As DependencyProperty = DependencyProperty.RegisterAttached("ExcludedData", GetType(IList), GetType(FilteredDataSupport), New PropertyMetadata(Nothing, AddressOf onDataChanged, AddressOf coerceExcludedData))

		Public Shared Sub SetExcludedData(ByVal element As UIElement, ByVal value As IList)
			element.SetValue(ExcludedDataProperty, value)
		End Sub
		Public Shared Function GetExcludedData(ByVal element As UIElement) As IList
			Return CType(element.GetValue(ExcludedDataProperty), IList)
		End Function

		Private Shared Sub onDataChanged(ByVal sender As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
			Dim grid As GridControl = TryCast(sender, GridControl)
			If grid Is Nothing Then
				Return
			End If
			If e.OldValue Is Nothing AndAlso e.NewValue IsNot Nothing Then
				If e.Property Is ExcludedDataProperty Then
					AddHandler grid.CustomRowFilter, AddressOf excludeOnCustomRowFilter
				ElseIf e.Property Is VisibleDataProperty Then
					AddHandler grid.CustomRowFilter, AddressOf OnCustomRowFilter
					AddHandler grid.FilterChanged, AddressOf OnFilterChanged
				End If
			ElseIf e.OldValue IsNot Nothing AndAlso e.NewValue Is Nothing Then
				If e.Property Is ExcludedDataProperty Then
					RemoveHandler grid.CustomRowFilter, AddressOf excludeOnCustomRowFilter
				ElseIf e.Property Is VisibleDataProperty Then
					RemoveHandler grid.CustomRowFilter, AddressOf OnCustomRowFilter
					RemoveHandler grid.FilterChanged, AddressOf OnFilterChanged
				End If
			End If
			grid.RefreshData()
		End Sub

		Private Shared Function coerceExcludedData(ByVal sender As DependencyObject, ByVal data As Object) As Object
			Dim grid As GridControl = TryCast(sender, GridControl)
			grid.RefreshData()
			Return data
		End Function

		Private Shared Sub excludeOnCustomRowFilter(ByVal sender As Object, ByVal e As RowFilterEventArgs)
			Dim grid As GridControl = TryCast(sender, GridControl)
			Dim data As IList = TryCast(grid.ItemsSource, IList)
			Dim excludedData As IList = TryCast(grid.GetValue(ExcludedDataProperty), IList)
			e.Visible = Not excludedData.Contains(data(e.ListSourceRowIndex))
			e.Handled = Not e.Visible
		End Sub
	End Class
End Namespace
